app/util/metrics.ts

```diff
@@ -0,0 +1,206 @@
+ */
+function constructBboxFromOperation(operation: DataOperation): BboxMetric {
+  const bbox = operation.boundingRectangle;
+  if (bbox) {
+    const [ west, south, east, north ] = bbox;
+    return { west, south, east, north };
+  }
+  return null;
+}
+
+/**
+ * Returns the request metric for a request
+ *
+ * @returns the request metric
+ */
+export function getRequestMetric(operation: DataOperation, serviceName: string): RequestMetric {
+  const rangeBeginDateTime = operation.temporal?.start;
+  const rangeEndDateTime = operation.temporal?.end;
+
+  const metric: RequestMetric = {
+    request_id: operation.requestId,
+    user_id: operation.user,
+    parameters: { service_name: serviceName },
+  };
+
+  const bbox = constructBboxFromOperation(operation);
+
+  if (bbox) {
+    metric.bbox = bbox;
+  }
+
+  if (rangeBeginDateTime) {
+    metric.rangeBeginDateTime = rangeBeginDateTime as string;
+  }
+
+  if (rangeEndDateTime) {
+    metric.rangeEndDateTime = rangeEndDateTime as string;
+  }
+
+  return metric;
+}
+
+/**
+ * Returns the product metric for a request
+ *
+ * @returns the product metric
+ */
+export function getProductMetric(operation: DataOperation, job: Job)
+  : ProductMetric {
+  let httpResponseCode = 200;
+
+  const failed = ([JobStatus.FAILED, JobStatus.CANCELED].includes(job.status));
+  if (failed) {
+    httpResponseCode = 500;
+  }
+
+  const regridding = operation.shouldReproject;
+  const subsetting =
+    operation.shouldVariableSubset ||
+    operation.shouldShapefileSubset ||
+    operation.shouldDimensionSubset ||
+    operation.shouldTemporalSubset ||
+    operation.shouldSpatialSubset;
+
+  const formatConversion = !!operation.outputFormat;
+  const subServices = {
+    regridding,
+    subsetting,
+    formatConversion,
+  };
+
+  const jobData = {
+    job_id: `${job.jobID}`,
+    startTime: job.createdAt.toISOString(),
+    endTime: job.updatedAt.toISOString(),
+    status: job.status,
+    sub_services: subServices,
+  };
+
+  const { sources } = operation;
+  const firstSource = sources[0];
+
+  const productData: ProductDataMetric = {
+    collectionId: firstSource.shortName + '___' + firstSource.versionId,
+    shortName: firstSource.shortName,
+    versionId: firstSource.versionId,
+  };
+
+  if (firstSource.variables) {
+    productData.variables = firstSource.variables.map((v) => v.name);
+  }
+
+  const metric = {
+    request_id: operation.requestId,
+    product_data: productData,
+    job_data: jobData,
+    http_response_code: httpResponseCode,
+  };
+
+  return metric;
+}
+
+/**
+ * Returns the response metric for a request
+ *
+ * @returns Promise that resolves to the response metric for a request
+ */
+export async function getResponseMetric(
+  operation: DataOperation, job: Job, originalSize: number,
+): Promise<ResponseMetric> {
+  let httpResponseCode = 200;
+
+  const failed = ([JobStatus.FAILED, JobStatus.CANCELED].includes(job.status));
+  if (failed) {
+    httpResponseCode = 500;
```

Comment:
Why 500? Guess I'm not really sure what `httpResponseCode` is supposed to reflect.

Reply:
I don't think it really makes sense for harmony except for synchronous requests. I was just using 500 to indicate failure and 200 to indicate success. I'll leave it this way for now, but we can always change it if people want something different in the future.

Reply:
It seems like we'll probably usually want it to be something else and to possibly have different error codes for different failure conditions, but I think this is okay for now